<a href="https://colab.research.google.com/github/R41YAN/Simulation-Modeling-Lab/blob/main/SIM_Lab2_Md_Raiyan_0432320005101101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Simulation and Modeling Lab  
 **Course Code:** CSE 413  


# Lab Assignment 2: Individualized Student Task - Unique Matrix Value Investigation

**Name:** Md. Raiyan  
**Student ID:** 0432320005101101  

**Objective:** To analyze how changing matrix values affects matrix properties (determinant, rank, eigenvalues, invertibility), using a student-specific matrix generated from my own ID.

## Necessary Libraries

In [45]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

## 1. Create My Matrix

My student ID is **0432320005101101**.

- Last digit of ID -> **d1 = 1**
- Second last digit of ID -> **d2 = 0**

$$A = \begin{bmatrix} d1+2 & d2+1 \\ 2d1 & d2+2 \end{bmatrix} = \begin{bmatrix} 3 & 1 \\ 2 & 2 \end{bmatrix}$$

In [46]:
# Build matrix A from my student ID digits
d1 = 1  # last digit of ID
d2 = 0  # second last digit of ID

A = np.array([[d1 + 2, d2 + 1],
              [2 * d1, d2 + 2]])

print("Matrix A:\n", A)

Matrix A:
 [[3 1]
 [2 2]]


## 2. Analyze Matrix A

In [47]:
shape_A = A.shape
det_A = np.linalg.det(A)
rank_A = np.linalg.matrix_rank(A)
eigenvalues_A = np.linalg.eigvals(A)

if det_A != 0:
    inv_A = np.linalg.inv(A)
else:
    inv_A = "Matrix is singular, cannot compute inverse"

print("Shape of A:", shape_A)
print("Determinant of A:", det_A)
print("Rank of A:", rank_A)
print("Eigenvalues of A:", eigenvalues_A)
print("Inverse of A:\n", inv_A)

Shape of A: (2, 2)
Determinant of A: 4.000000000000001
Rank of A: 2
Eigenvalues of A: [4. 1.]
Inverse of A:
 [[ 0.5  -0.25]
 [-0.5   0.75]]


## 3. Value Change Experiment

Created matrix **B** by changing exactly one value in A: Added **+1** to the element at position **A[0][0]** (row 1, column 1), which changes it from 3 to 4.

$$B = \begin{bmatrix} 4 & 1 \\ 2 & 2 \end{bmatrix}$$

In [48]:
# Create matrix B by adding 1 to element A[0][0]
B = A.copy()

# Add +1 to the element at row 1, column 1
B[0, 0] += 1

print("Matrix A:")
print(A)

print("\nMatrix B:\n", B)

Matrix A:
[[3 1]
 [2 2]]

Matrix B:
 [[4 1]
 [2 2]]


## 4. Re-analyze Matrix B

In [50]:
shape_B = B.shape
det_B = np.linalg.det(B)
rank_B = np.linalg.matrix_rank(B)
eigenvalues_B = np.linalg.eigvals(B)

if det_B != 0:
    inv_B = np.linalg.inv(B)
else:
    inv_B = "Matrix is singular, cannot compute inverse"

print("Shape of B:", shape_B)
print("Determinant of B:", det_B)
print("Rank of B:", rank_B)
print("Eigenvalues of B:", eigenvalues_B)
print("Inverse of B:\n", inv_B)

Shape of B: (2, 2)
Determinant of B: 6.0
Rank of B: 2
Eigenvalues of B: [4.73205081 1.26794919]
Inverse of B:
 [[ 0.33333333 -0.16666667]
 [-0.33333333  0.66666667]]


## 5. Comparison & Explanation

**1. How did the determinant change and why?**

=> The determinant increased from 4 to 6. I added 1 to A[0][0] (3→4), and since det = a*d - b*c, a bigger `a` with a positive `d` (=2) directly increases the determinant by 1×2 = 2.

**2. Did the rank change? Explain.**

=> No, rank stayed 2 for both. Both determinants are non-zero, so both matrices are still full rank — a small +1 change wasn't enough to make the rows dependent.

**3. How did the eigenvalues respond to the value change?**

=> Both eigenvalues went up: A had 4 and 1, B had about 4.73 and 1.27. This tracks with the trace (sum of diagonal elements), which rose from 5 to 6, so the eigenvalues had to rise too since they sum to the trace.

**4. Is B easier or harder to invert than A? Why?**

=> Both are easily invertible since neither determinant is zero. B is slightly "safer" to invert since its determinant (6) is larger than A's (4), meaning it's further from being singular.